server: introducing DDoS prevention mechanisms

chemicL opened the pull request:
As described in #101 – I'm suggesting adding a DDoS prevention layer and this is an attempt to address this issue.

The work done here is partly based on the example implementation of [manual flow control in grpc-java](https://github.com/grpc/grpc-java/blob/master/examples/src/main/java/io/grpc/examples/manualflowcontrol/ManualFlowControlServer.java).

I tried making everything customizable and pluggable. The existing usages of `java-control-plane` should not be affected by these changes and will be available upon using classes in the`limits` package as depicted in `TestLimits` runner.

To perform load tests I used [ghz](https://github.com/bojand/ghz). Here is the configuration and a run command:

```
{
  "proto": "proto/cds.proto",
  "importPaths": "proto",
  "call": "envoy.api.v2.ClusterDiscoveryService/StreamClusters",
  "n": 45,
  "c": 15,
  "t": 60,
  "host": "0.0.0.0:12345",
  "insecure": true,
  "d": {
    "node": {
      "id": "myself",
      "cluster": "somecluster",
      "metadata": {},
      "locality": {
        "region": "unknown",
        "zone": "one",
        "sub_zone": ""
      }
    },
    "resource_names": [],
    "response_nonce": "",
    "type_url": "type.googleapis.com/envoy.api.v2.Cluster"
  }
}
```

```ghz -config java-control-plane-loadtest.json```

To run this, I copied over the `proto` folder into my testing directory.

What this configuration does:

- Opens 45 streams,
- Concurrently sends 15 requests,
- Times out a stream after 60 seconds,
- For each stream it sends an initial CDS request.

With the `TestLimits` configuration of

- 1 RPS rate limit,
- 10 concurrent streams.

This test takes around 30s to finish and results in 15 streams not being handled.

ghz's output:

```
Summary:
  Count:	45
  Total:	28.01 s
  Slowest:	10.00 s
  Fastest:	6.27 ms
  Average:	5.23 s
  Requests/sec:	1.61

Response time histogram:
  6.272 [1]	|∎∎
  1006.015 [1]	|∎∎
  2005.759 [1]	|∎∎
  3005.502 [1]	|∎∎
  4005.245 [1]	|∎∎
  5004.988 [1]	|∎∎
  6004.731 [1]	|∎∎
  7004.474 [1]	|∎∎
  8004.218 [1]	|∎∎
  9003.961 [2]	|∎∎∎∎
  10003.704 [19]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎

Latency distribution:
  10% in 2.01 s
  25% in 7.01 s
  50% in 10.00 s
  75% in 10.00 s
  90% in 10.00 s
  95% in 10.00 s
  0% in 0 ns
Status code distribution:
  [OK]            30 responses
  [Unavailable]   15 responses

Error distribution:
  [15]   rpc error: code = Unavailable desc =
```

Regarding code coverage. I tried doing my best, but due to the nature of `ServerCallStreamObserver` and the hidden default implementation, mocking it would miss the point in my opinion. When it comes to the part of limiting concurrently open streams, I tried depicting it in `TestLimits` runner and providing a load test configuration. I'm open to suggestions if you have better ideas for automated tests that I can add.

Fixes #101

snowp commented on server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java:
```diff
@@ -339,4 +377,79 @@ private void send(Response response, String typeUrl) {
+     */
+    public DiscoveryServer build() {
+      List<DiscoveryServerCallbacks> callbacks = this.callbacks == null
+          ? Collections.emptyList()
```
You can initialize the fields with emptyList instead of using these ternaries, and do similar default initialization of the other args

chemicL replied:
I wanted to avoid unnecessary initialization of these objects, but they cost close to nothing, I'll do as you suggest.

snowp commented on server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java:
```diff
@@ -339,4 +377,79 @@ private void send(Response response, String typeUrl) {
+    }
+
+    public Builder withCallbacks(List<DiscoveryServerCallbacks> callbacks) {
+      this.callbacks = callbacks;
```
i'd do a `Preconditions.checkNotNull` on all these

snowp commented on server/src/main/java/io/envoyproxy/controlplane/server/limits/FlowControl.java:
```diff
@@ -0,0 +1,23 @@
+
+import io.grpc.stub.StreamObserver;
+
+public interface FlowControl<V> {
```
Mind adding javadocs? Here and on other public classes

snowp commented on server/src/main/java/io/envoyproxy/controlplane/server/limits/ManualFlowControl.java:
```diff
@@ -0,0 +1,78 @@
+public class ManualFlowControl implements FlowControl<DiscoveryResponse> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ManualFlowControl.class);
+
```
nit: other code in this codebase avoids newlines between fields

chemicL replied:
Sort of yes, but not consistently :-) Check SimpleCache for example. Of course I'll remove newlines.

snowp commented on server/src/main/java/io/envoyproxy/controlplane/server/limits/ManualFlowControl.java:
```diff
@@ -0,0 +1,78 @@
+    serverCallStreamObserver.setOnReadyHandler(() -> {
+      if (serverCallStreamObserver.isReady() && observerWasReady.compareAndSet(false, true)) {
+        LOGGER.debug("[{}] stream ready", streamId);
+        serverCallStreamObserver.request(1);
```
Should you be acquiring a rate limit here? Conceptually it seems like each permit maps to a request?

chemicL replied:
Well, this is the situation when the consumer's (client reading responses) HTTP2 buffer gains enough space to consume more messages. In this case it's not really a message **yet**. The permit is performed **after** processing the request, not before.

snowp commented on server/src/main/java/io/envoyproxy/controlplane/server/limits/StreamLimiter.java:
```diff
@@ -0,0 +1,85 @@
+
+          @Override
+          public void onHalfClose() {
+            delegate.onHalfClose();
```
I'm hazy on these callbacks: does a half close precede a onCancel/onComplete? Do you need to decrement the counter here?

chemicL replied:
I was a bit confused too TBH so I checked the docs:
```<p>Any contexts are guaranteed to arrive before any messages, which are guaranteed before half close, which is guaranteed before completion.``` (https://github.com/grpc/grpc-java/blob/80c3c992a66aa21ccf3e12e38000316e45f97e64/api/src/main/java/io/grpc/ServerCall.java#L45)

snowp commented on server/src/main/java/io/envoyproxy/controlplane/server/limits/StreamLimiter.java:
```diff
@@ -0,0 +1,85 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class StreamLimiter {
```
Given that this is just a generic gRPC stream limiting interceptor, does it make sense to commit this to this repo? Seems like if it doesn't actually tie in with the control plane code itself it might just add to the maintenance burden without a clear benefit?

Maybe @joeyb has thoughts here?

ben-hoyt replied:
+1. Related: https://github.com/Netflix/concurrency-limits/tree/master/concurrency-limits-grpc

chemicL replied:
I'm not feeling extremely strong about it, but it seems more feature complete if it resides in this repository. To provide a full example such as the one in `TestLimits` would require importing the external dependency from Netflix in the test package.

chemicL commented on server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java:
```diff
@@ -236,7 +270,8 @@ public void onNext(DiscoveryRequest request) {
       }
 
-      LOGGER.info("[{}] request {}[{}] with nonce {} from version {}",
+      LOGGER.info(
```
I did follow https://github.com/envoyproxy/java-control-plane/blob/master/CONTRIBUTING.md and am using the google-java-format plugin and the style mentioned. I had to manually revert a lot of automatically converted changes not related to my additions and must have missed this one. The google style guide mentions 100 as the limit and that's what the plugin enforces. I'll update my settings to reflect this, but it would be appreciated if it was mentioned in the docs; or just formatted according to the plugin :-)

chemicL commented on server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java:
```diff
@@ -296,6 +333,7 @@ public void onCompleted() {
 
       try {
         callbacks.forEach(cb -> cb.onStreamClose(streamId, defaultTypeUrl));
+        flowControl.streamClosed();
```
True, I'll move these to finally block.

chemicL commented:
@snowp thanks for having a look and for your remarks. I addressed your requests and updated the PR. Please do let me know what's your view on the server interceptor and if anything else needs changes.

chemicL commented on pom.xml:
```diff
@@ -40,7 +40,7 @@
         <maven-failsafe-plugin.version>2.21.0</maven-failsafe-plugin.version>
         <maven-gpg-plugin.version>1.6</maven-gpg-plugin.version>
-        <maven-javadoc-plugin.version>3.0.0</maven-javadoc-plugin.version>
+        <maven-javadoc-plugin.version>3.1.0</maven-javadoc-plugin.version>
```
I had to update the plugin, as I was hitting this: https://issues.apache.org/jira/browse/LANG-1365

chemicL commented:
I just rebased with master to resolve conflicts in imports.

chemicL commented:
I'm closing this PR, more details here:
https://github.com/envoyproxy/java-control-plane/issues/101#issuecomment-527449648
